pkg/asset/installconfig/aws/validation.go

```diff
@@ -0,0 +1,162 @@
+	allErrs = append(allErrs, validateDuplicateSubnetZones(fldPath, privateSubnets, privateSubnetsIdx, "private")...)
+	allErrs = append(allErrs, validateDuplicateSubnetZones(fldPath, publicSubnets, publicSubnetsIdx, "public")...)
+
+	privateZones := sets.NewString()
+	publicZones := sets.NewString()
+	for _, subnet := range privateSubnets {
+		privateZones.Insert(subnet.Zone)
+	}
+	for _, subnet := range publicSubnets {
+		publicZones.Insert(subnet.Zone)
+	}
+	if publish == types.ExternalPublishingStrategy && !publicZones.IsSuperset(privateZones) {
+		errMsg := fmt.Sprintf("No public subnet provided for zones %s", privateZones.Difference(publicZones).List())
+		allErrs = append(allErrs, field.Invalid(fldPath, subnets, errMsg))
+	}
+
+	return allErrs
+}
+
+func validateMachinePool(ctx context.Context, meta *Metadata, fldPath *field.Path, platform *awstypes.Platform, pool *awstypes.MachinePool) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if len(pool.Zones) > 0 {
+		availableZones := sets.String{}
+		if len(platform.Subnets) > 0 {
+			privateSubnets, err := meta.PrivateSubnets(ctx)
+			if err != nil {
+				return append(allErrs, field.InternalError(fldPath, err))
+			}
+			for _, subnet := range privateSubnets {
+				availableZones.Insert(subnet.Zone)
+			}
+		} else {
+			allzones, err := meta.AvailabilityZones(ctx)
+			if err != nil {
+				return append(allErrs, field.InternalError(fldPath, err))
+			}
+			availableZones.Insert(allzones...)
+		}
+
+		if diff := sets.NewString(pool.Zones...).Difference(availableZones); diff.Len() > 0 {
+			errMsg := fmt.Sprintf("No subnets provided for zones %s", diff.List())
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("zones"), pool.Zones, errMsg))
+		}
+	}
+	return allErrs
+}
+
+func validateSubnetCIDR(fldPath *field.Path, subnets map[string]Subnet, idxMap map[string]int, machineCIDR *ipnet.IPNet) field.ErrorList {
+	allErrs := field.ErrorList{}
+	for id, v := range subnets {
+		fp := fldPath.Index(idxMap[id])
+		cidr, _, err := net.ParseCIDR(v.CIDR)
+		if err != nil {
+			allErrs = append(allErrs, field.Invalid(fp, id, err.Error()))
+			continue
+		}
+		if !machineCIDR.Contains(cidr) {
+			errMsg := fmt.Sprintf("CIDR range %s is outside of the MachineCIDR %s", v.CIDR, machineCIDR)
+			allErrs = append(allErrs, field.Invalid(fp, id, errMsg))
+			continue
+		}
+	}
+	return allErrs
+}
+
+func validateDuplicateSubnetZones(fldPath *field.Path, subnets map[string]Subnet, idxMap map[string]int, typ string) field.ErrorList {
+	var keys []string
+	for id := range subnets {
+		keys = append(keys, id)
+	}
+	sort.Strings(keys)
+
+	allErrs := field.ErrorList{}
+	zones := map[string]string{}
+	for _, id := range keys {
+		subnet := subnets[id]
+		if conflictingSubnet, ok := zones[subnet.Zone]; ok {
+			errMsg := fmt.Sprintf("%s subnet %s is also in zone %s", typ, conflictingSubnet, subnet.Zone)
```

Comment:
nit: not sure but I think a more helpful error message would be of the form: 
`zone a may only contain one private subnet but contaIns multiple: foo, bar`

right now the error message would be something like: 
`private subnet bar is also in zone a`

but I could be missing something re aws subnets.

Reply:
The current validation error is fine IMO, and we get issues that it needs clarity, we can fix this in follow-up